packages/ra-core/src/auth/useHandleAuthCallback.ts

```diff
@@ -24,15 +24,38 @@ export const useHandleAuthCallback = (
     const nextSearch = locationState && locationState.nextSearch;
     const defaultRedirectUrl = nextPathName ? nextPathName + nextSearch : '/';
     const { onSuccess, onError, onSettled, ...queryOptions } = options ?? {};
+    let handleCallbackPromise: Promise<void> | null;
 
     const queryResult = useQuery({
         queryKey: ['auth', 'handleCallback'],
-        queryFn: ({ signal }) =>
-            authProvider && typeof authProvider.handleCallback === 'function'
-                ? authProvider
-                      .handleCallback({ signal })
-                      .then(result => result ?? null)
-                : Promise.resolve(),
+        queryFn: ({ signal }) => {
+            console.log('queryFn', handleCallbackPromise);
+            if (!handleCallbackPromise) {
+                handleCallbackPromise = new Promise(async (resolve, reject) => {
+                    if (authProvider) {
+                        if (typeof authProvider.handleCallback === 'function') {
+                            try {
+                                const result =
+                                    await authProvider.handleCallback({
+                                        signal,
+                                    });
+                                return resolve(result ?? null);
+                            } catch (error) {
+                                return reject({
+                                    redirectTo: false,
+                                    message: error.message,
+                                });
+                            }
+                        }
+                        return resolve();
+                    }
+                    return reject({
+                        message: 'Failed to handle login callback.',
+                    });
+                });
```

Comment:
This changes the previous behavior. Please revert to the original code. Only keep the `handleCallbackPromise` creation as you did